maven-builder-plugin/src/main/java/hudson/tasks/Maven.java

```diff
@@ -733,7 +733,9 @@ public boolean isApplicable(Class<? extends ToolInstallation> toolType) {
      * to use the consistent Maven installation attached to the project.
      *
      * @since 1.235
+     * @deprecated Use WithMaven extension point
      */
+    @Deprecated
     public interface ProjectWithMaven {
```

Comment:
Now you are incompatible for existing implementations of `ProjectWithMaven`, since you are not checking for it in `obtainMavenInstallation`.

Would it not be simpler to just delete `MavenSelector` and keep `ProjectWithMaven` as is, merely adding a utility method something like

```java
public @CheckForNull MavenInstallation inferMavenInstallation(@NonNull Job<?,?> job) {
    if (job instanceof ProjectWithMaven) {
        return ((ProjectWithMaven) job).inferMavenInstallation();
    } else if (job instanceof Project) {
        Maven m = ((Project<?, ?>) item).getBuildersList().get(Maven.class);
            if (m != null) {
                return m.getMaven();
            }
        }
    }
    return null;
}
````

? Then the only patch you need for `maven-plugin` would be to call the new utility method (rather than checkcasting to `ProjectWithMaven`) in `MavenUtil.createEmbedder`. Plugins which merely implemented `ProjectWithMaven` need not be touched at all (`DETACHED_PLUGINS` will make them work).